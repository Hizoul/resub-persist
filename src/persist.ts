import * as get from "lodash.get"
import * as isArray from "lodash.isarray"
import * as set from "lodash.set"
import makeObjectFromPropKeys from "./makeObjectFromPropKeys"
import { IPersistableStore } from "./type"

const persist = (storage: any, stores: IPersistableStore[]) => {
  return new Promise((resolve) => {
    let neededCallbacks = isArray(stores) ? stores.length : 0
    const callb = () => {
      neededCallbacks--
      if (neededCallbacks === 0) {
        resolve()
      }
    }
    for (const store of stores) {
      storage.setItem(`store.${store.name}`, JSON.stringify(makeObjectFromPropKeys(store)), callb)
    }
  })
}

export default persist
