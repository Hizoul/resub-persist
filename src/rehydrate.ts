import * as get from "lodash.get"
import * as set from "lodash.set"
import { IPersistableStore, KeyOrKeys } from "./type"

const rehydrate = (storage: any, stores: IPersistableStore[]) => {
  return new Promise((resolve) => {
    let callbacksNeeded = stores.length
    const tryResolve = () => {
      callbacksNeeded--
      if (callbacksNeeded === 0) {
        resolve()
      }
    }
    for (const store of stores) {
      storage.getItem(`store.${store.name}`, (err: any, objString: any) => {
        const obj = JSON.parse(objString)
        for (const key of store.getPropKeys()) {
          const oldValue = get(obj, key)
          if (oldValue !== undefined && oldValue !== null) {
            set(store, key, oldValue)
          }
        }
        const untypedStore: any = store
        untypedStore.trigger(store.rehydratedKeys)
        tryResolve()
      })
    }
  })
}

export default rehydrate
