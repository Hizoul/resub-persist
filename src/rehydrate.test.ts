import persist from "./persist"
import rehydrate from "./rehydrate"
import memForage from "./testUtil/miniMemForage"
import TestStore, { TestStore1 } from "./testUtil/testStore"

test("Cover Persist Proeprty Test", (done) => {
  const secondaryStore: any = new TestStore1()
  TestStore.toggleMyBool()
  TestStore.setMyString("mychangedString")
  expect(memForage.getData()).toMatchSnapshot("Before Persistence Storage is Empty")
  persist(memForage, [TestStore])
  expect(memForage.getData()).toMatchSnapshot("After Persistence Storage is contains changes")
  expect(TestStore.isMyBool()).not.toBe(secondaryStore.isMyBool())
  expect(TestStore.isMyString()).not.toBe(secondaryStore.isMyString())
  rehydrate(memForage, [secondaryStore]).then(() => {
    expect(TestStore.isMyBool()).toBe(secondaryStore.isMyBool())
    expect(TestStore.isMyString()).toBe(secondaryStore.isMyString())
    done()
  })
})
