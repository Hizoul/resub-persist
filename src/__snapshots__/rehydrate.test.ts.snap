// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cover Persist Proeprty Test: After Persistence Storage is contains changes 1`] = `
Object {
  "store.detailList": "{\\"myBool\\":true,\\"myString\\":\\"mychangedString\\"}",
}
`;

exports[`Cover Persist Proeprty Test: Before Persistence Storage is Empty 1`] = `Object {}`;
